modules/lang-painless/src/main/java/org/elasticsearch/painless/AnalyzerCaster.java

```diff
@@ -30,6 +30,9 @@
 public final class AnalyzerCaster {
 
     public static Cast getLegalCast(Location location, Type actual, Type expected, boolean explicit, boolean internal) {
+        if (actual == null || expected == null) {
```

Comment:
I figure if I end up scratching my head for a while on an NPE it is worth being specific about the message.

The tests that I added about writing to a map triggered this. It is why I have a guard on the `cast` call [here](https://github.com/elastic/elasticsearch/pull/21239/files#diff-d9a6383229f3c2e2265023dd67f55ae1R57). I'm spent a while digging but it looks like it is intentional that `expected` can be null in some cases and I have to not try and cast.